cql3/expr/prepare_expr.cc

```diff
@@ -986,6 +986,12 @@ try_prepare_expression(const expression& expr, data_dictionary::database db, con
             }
             auto& schema = *schema_opt;
 
+            if (receiver.get() != nullptr && &receiver->type->without_reversed() != long_type.get()) {
+                throw exceptions::invalid_request_exception(
+                    format("token() produces a value fo type long, which doesn't match the type: {} of {}",
+                           receiver->type->name(), receiver->name->text()));
+            }
```

Comment:
We should have something like require_receiver_type(...) to avoid this boilerplate. Shouldn't block this series.

Reply:
Also, one day we'll need to handle integer promotions (so we shouldn't check if the types are exactly equal, just if they are compatible), but no need to rush with that. Most type inference is for guessing the types of bind variables.